data_model/derive/src/id.rs

```diff
@@ -1,120 +1,121 @@
             }
         }
 
-        impl core::cmp::PartialEq for #name {
+        impl #impl_generics core::cmp::Eq for #name #ty_generics #where_clause where Self: Identifiable  {}
+        impl #impl_generics core::cmp::PartialEq for #name #ty_generics #where_clause  where Self: Identifiable {
             fn eq(&self, other: &Self) -> bool {
                 self.id() == other.id()
             }
         }
 
-        impl core::cmp::Eq for #name {}
-
-        impl core::hash::Hash for #name {
+        impl #impl_generics core::hash::Hash for #name #ty_generics #where_clause  where Self: Identifiable {
             fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
                 self.id().hash(state);
             }
         }
     }
 }
 
-pub fn impl_id(ast: &IdInput) -> TokenStream {
-    let id = &ast.id_type;
-    let name = &ast.ident;
+fn get_id_type(input: &syn::ItemStruct) -> (TokenStream, TokenStream) {
+    match &input.fields {
+        syn::Fields::Named(fields) => {
+            for field in &fields.named {
+                let (field_name, field_ty) = (&field.ident, &field.ty);
 
-    let ordeqhash = impl_ordeqhash(ast);
+                if is_identifier(&field.attrs) {
+                    return (quote! {#field_ty}, quote! {&self.#field_name});
+                }
+                if is_transparent(&field.attrs) {
+                    return (
+                        quote! {<#field_ty as Identifiable>::Id},
+                        quote! {Identifiable::id(&self.#field_name)},
+                    );
+                }
+            }
+        }
+        syn::Fields::Unnamed(fields) => {
+            for (i, field) in fields.unnamed.iter().enumerate() {
+                let (field_id, field_ty): (syn::Index, _) = (i.into(), &field.ty);
 
-    let body = if ast.ident.to_string().starts_with("NewRole") {
-        // `NewRole` struct only has unconventional body
-        quote! { &self.inner.id }
-    } else {
-        // Most usual case for many `data_model` structs
-        quote! { &self.id }
-    };
-    quote! {
-        impl Identifiable for #name {
-            type Id = #id;
+                if is_identifier(&field.attrs) {
+                    return (quote! {#field_ty}, quote! {&self.#field_id});
+                }
+                if is_transparent(&field.attrs) {
+                    return (
+                        quote! {<#field_ty as Identifiable>::Id},
+                        quote! {Identifiable::id(&self.#field_id)},
+                    );
+                }
+            }
+        }
+        syn::Fields::Unit => {}
+    }
 
-            #[inline]
-            fn id(&self) -> &Self::Id {
-                #body
+    match &input.fields {
+        syn::Fields::Named(named) => {
+            for field in &named.named {
+                let field_ty = &field.ty;
+
+                #[allow(clippy::expect_used)]
+                if field.ident.as_ref().expect("Field must be named") == "id" {
```

Comment:
Is panicking better than `abort!` here?

Reply:
this literally can never happen. I could have used `unwrap`

Reply:
Just because something is prevented by an invariant doesn't mean that it can't happen.

Reply:
but we agree this is ok?